File: 0001-find-pair-with-given-sum-in-the-array/03-hashing/main.java
import java.util.HashMap;
import java.util.Map;

class FindPair
{
    // Naive method to find a pair in an array with given sum
    public static void findPair(int arr[], int sum)
    {
        // Create an empty Hash Map
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();

        // do for each element
        for (int i = 0; i < arr.length; i++)
            {
                // check if pair (arr[i], sum-arr[i] exists

                // if difference is seen before, print the pair
                if (map.containsKey(sum - arr[i]))
                    {
                        System.out.println("Pair found at index " +
                                           map.get(sum - arr[i]) + " and " + i);
                        return;
                    }

                // store the index of current element in the map
                map.put(arr[i], i);
            }

        // No pair with given sum exists in the array
        System.out.println("Pair not found");
    }

    // main function
    public static void main (String[] args)
    {
        int arr[] = { 8, 7, 2, 5, 3, 1 };
        int sum = 10;

        findPair(arr, sum);
    }
}

File: 0001-find-pair-with-given-sum-in-the-array/03-hashing/main.py
import sys

def findPair(arr, k):
    cache = {}

    for i in range(0, len(arr)):
        if arr[i] in cache.keys():
            print "Pair found at index %s and %s" % (cache[arr[i]], i)
            sys.exit(0)

        cache[k-arr[i]] = i

    print "Pair not found"
    print cache.keys()


if __name__ == '__main__':
    arr = [ 8, 7, 2, 5, 3, 1 ]
    k = 10

    findPair(arr, k)
